libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -467,56 +469,42 @@ static inline int vgic_add_overflow_cpu(struct lr_of *lr_overflow, struct virq_h
     return 0;
 }
 
-static inline int vgic_add_overflow(vgic_t *vgic, struct virq_handle *irq, vm_vcpu_t *vcpu)
+static inline void vgic_handle_overflow(vgic_t *vgic, vm_vcpu_t *vcpu)
 {
-    return vgic_add_overflow_cpu(&vgic->lr_overflow[vcpu->vcpu_id], irq);
-}
+    struct lr_of *lr_overflow = &vgic->lr_overflow[vcpu->vcpu_id];
 
-static inline void vgic_handle_overflow_cpu(vgic_t *vgic, struct lr_of *lr_overflow, vm_vcpu_t *vcpu)
-{
-    /* copy tail, as vgic_vcpu_inject_irq can mutate it, and we do
-     * not want to process any new overflow irqs */
-    size_t tail = lr_overflow->tail;
-    for (size_t i = lr_overflow->head; i != tail; i = LR_OF_NEXT(i)) {
-        if (vgic_vcpu_inject_irq(vgic_dist, vcpu, &lr_overflow->irqs[i]) == 0) {
-            lr_overflow->head = LR_OF_NEXT(i);
-            lr_overflow->full = (lr_overflow->head == LR_OF_NEXT(lr_overflow->tail));
-        } else {
+    for (size_t i = lr_overflow->head; i != lr_overflow->tail; i = LR_OF_NEXT(i)) {
+        int err = vgic_vcpu_load_lr(vgic_dist, vcpu, &lr_overflow->irqs[i]);
+        if (err) {
             break;
         }
+        lr_overflow->head = LR_OF_NEXT(i);
+        lr_overflow->full = (lr_overflow->head == LR_OF_NEXT(lr_overflow->tail));
```

Comment:
Can we simply say
```suggestion
        lr_overflow->full = false; /* at least 1 slot is free now */
```
here, because when we removed an element from the queue, this implies the queue is no longer full.

Reply:
Yes.